### PR TITLE
Update to Scintilla 5.1.4 and Lexilla 5.1.3

### PR DESCRIPTION
**Pull request by kugel-**
Notable changes:
Scintilla:
- Add SC_ELEMENT_FOLD_LINE to set the colour of fold lines. Add SC_ELEMENT_HIDDEN_LINE to show where lines are hidden.
- On GTK, fix the line spacing so that underscores and accents are visible for some fonts such as DejaVu Sans Mono 10.
- On GTK, fix primary selection paste within same instance. Bug https://sourceforge.net/p/scintilla/bugs/2287/
- On GTK, fix potential crash when closing Scintilla instances due to releasing global settings object.

Lexilla
- Implement conditional group rules in CSS. Issue https://github.com/ScintillaOrg/lexilla/issues/25, Pull request https://github.com/ScintillaOrg/lexilla/pull/28.
- Check PHP numeric literals, showing invalid values with default style instead of numeric. Issue https://github.com/ScintillaOrg/lexilla/issues/20.
- Various Rust fixes

The CSS change causes compatiblity trouble. We exposed the changed style
in filetypes.css. We hope that not too many users have a local filetypes.css
because they must update them.

[Edited PR/Issue numbers linked to lexilla repo, not this repo]

### CODE REVIEW

File: scintilla/gtk/ScintillaGTK.cxx [hunk 1]
```diff
@@ -179,6 +180,18 @@ GdkAtom SelectionOfGSD(GtkSelectionData *sd) noexcept {
 	return gtk_selection_data_get_selection(sd);
 }
 
+bool SettingGet(GtkSettings *settings, const gchar *name, gpointer value) noexcept {
```
Comment (elextr) [inline scintilla/gtk/ScintillaGTK.cxx:183]: This looks like the only change that is likely to cause segfaults on widget destroy of the main window if these settings are added to the Scintilla widget and the memory is double released?

File: scintilla/gtk/ScintillaGTK.cxx [hunk 2]
```diff
@@ -264,6 +279,12 @@ ScintillaGTK::~ScintillaGTK() {
 	}
 	ClearPrimarySelection();
 	wPreedit.Destroy();
+	if (settingsHandlerId) {
+		g_signal_handler_disconnect(settings, settingsHandlerId);
+	}
+	if (settings) {
+		g_object_unref(settings);
```
Comment (kugel-) [inline scintilla/gtk/ScintillaGTK.cxx:286]: Fairly sure this is the problem. `settings` is initialized using `gtk_settings_get_default()` and this is documented as "transfer-none", so `g_object_unref()` is inappropriate. Code using this instance is then suffering from use-after-free.

Comment (kugel-) [reply scintilla/gtk/ScintillaGTK.cxx:286]: https://sourceforge.net/p/scintilla/code/ci/db178448ae7a1049853c8e6e2cdba6f798ec1ed5/

I guess we can simply remove the unref as the same fix is alredy applied upstream.

File: scintilla/gtk/ScintillaGTK.cxx [hunk 3]
```diff
@@ -282,9 +282,6 @@ ScintillaGTK::~ScintillaGTK() {
 	if (settingsHandlerId) {
 		g_signal_handler_disconnect(settings, settingsHandlerId);
 	}
-	if (settings) {
```
Comment (elextr) [inline scintilla/gtk/ScintillaGTK.cxx]: Fixes crash here.

**Comment by elextr** [comment]
Has been working for me for two weeks (with some but not heavy testing), just needs the second CSS GROUP_RULE line removed, it does not work.  

So we have to accept that there is a breakage of user edited filetypes.css but thats going to happen any time Lexilla changes syntactic entities incompatibility.  I do not think Geany should start committing to a backward compatibility guarantee that Lexilla does not provide.

**Comment by xiota** [comment]
It's reasonable to expect users who edit `filetypes.css` to check the manual if anything breaks because the first three lines state:
```
# For complete documentation of this file, please see Geany's main documentation
[styling]
# Edit these in the colorscheme .conf file instead
```

**Comment by kugel-** [comment]
@elextr good to hear we're on the same page.

I can update this PR but I'm unsure if we should wait for 5.1.4 ?

**Comment by elextr** [comment]
@kugel- Neil just posted 5.1.4 will be "in a few days" so it probably better to wait.

**Comment by kugel-** [comment]
updated scintilla and removed the defunct compatiblity change.

**Comment by elextr** [comment]
@kugel- thanks, will test

**Comment by xiota** [comment]
I've rebased and built this off of master...  nothing obvious has gone wrong so far...  is there anything I should look for?

**Comment by elextr** [comment]
@xiota Anything mentioned in the comments above as a problem with previous Scintilla 5 versions, then look for any new highlighting problems in any language you know/use.  Then just use it, thats usually when things appear.

**Comment by elextr** [comment]
So far I have found nothing, will keep using it for the rest of the week.  It would be nice if @eth16 and @techee (or anybody else who can build and run on those platforms of course) could try on windows and mac.

**Comment by eht16** [comment]
> So far I have found nothing, will keep using it for the rest of the week. It would be nice if @eht16 [...] could try on windows and mac.

Will do but it will take until the end of November. IMO it'd be OK to merge into master even if Windows is not yet tested. Don't let it getting old here more than necessary.

**Comment by kugel-** [comment]
It's not a major update since 5.1.1 is already in master so I don't expect wildness on Windows. I can have a look too, though.

**Comment by xiota** [comment]
> Don't let it getting old here more than necessary.

 Update to Scintilla 5.1.5 and Lexilla 5.1.4 ...

**Comment by elextr** [comment]
> Don't let it getting old here more than necessary.

Yeah, its "nice" not necessary that other platforms are checked before merge, we are a ways off any release, it will get tested plenty before then with split session and ctags stuff to do as well as billions [exaggeration for effect] of small paper cuts.

**Comment by kugel-** [comment]
I swiftly checked windows and it looks good (opened a C file and .ini file, nothing in-depth)

**Comment by kugel-** [comment]
Should the filetypes.css thing be noted anywhere?

**Comment by elextr** [comment]
The CSS thing is a temporary thing for one version, so I don't think its anything that needs to be in the manual.  Probably should be in the release notes for the next release, if we remember it.

**Comment by kugel-** [comment]
What do you mean with "for one version"? Don't we have filetype.css for a long time?

**Comment by elextr** [comment]
What I mean is that the incompatibility is between two versions, this version may be incompatible with the 1.38 but (hopefully) the next version will be compatible with this one.  Geany doesn't document syntactic entities anyway, so its just a warning at release.

**Comment by techee** [comment]
> So far I have found nothing, will keep using it for the rest of the week. It would be nice if @eht16 and @techee (or anybody else who can build and run on those platforms of course) could try on windows and mac.

I haven't tried this specific pull request but update to Scintilla 5 went without any Mac-specific problems so I don't expect any problem here.

**Pull request merged**